channeldb/migration26/iterator.go

```diff
@@ -0,0 +1,257 @@
+// returns the commit height bytes found. A nil return value means the
+// migration has been completed for this particular channel bucket.
+func findNextMigrateHeight(chanBucket kvdb.RBucket) []byte {
+	// Read the old log bucket. The old bucket doesn't exist, indicating
+	// either we don't have any old logs for this channel, or the migration
+	// has been finished and the old bucket has been deleted.
+	oldBucket := chanBucket.NestedReadBucket(
+		revocationLogBucketDeprecated,
+	)
+	if oldBucket == nil {
+		return nil
+	}
+
+	// Acquire a read cursor for the old bucket.
+	oldCursor := oldBucket.ReadCursor()
+
+	// Read the new log bucket. The sub-bucket hasn't been created yet,
+	// indicating we haven't migrated any logs under this channel. In this
+	// case, we'll return the first commit height found from the old
+	// revocation log bucket as the next height.
+	logBucket := chanBucket.NestedReadBucket(revocationLogBucket)
+	if logBucket == nil {
+		nextHeight, _ := oldCursor.First()
+		return nextHeight
+	}
+
+	// Acquire a read cursor for the new bucket.
+	cursor := logBucket.ReadCursor()
+
+	// Read the last migrated record. If the key is nil, we haven't
+	// migrated any logs yet. In this case we return the first commit
+	// height found from the old revocation log bucket.
+	migratedHeight, _ := cursor.Last()
+	if migratedHeight == nil {
+		nextHeight, _ := oldCursor.First()
+		return nextHeight
+	}
+
+	// Read the last height from the old log bucket. If the height of the
+	// last old revocation equals to the migrated height, we've done
+	// migrating for this channel.
+	endHeight, _ := oldCursor.Last()
+	if bytes.Equal(migratedHeight, endHeight) {
+		return nil
+	}
+
+	// Now point the cursor to the migratedHeight. If we cannot find this
+	// key from the old log bucket, the database might be corrupted. In
+	// this case, we would return the first key so that we would redo the
+	// migration for this chan bucket.
+	matchedHeight, _ := oldCursor.Seek(migratedHeight)
+	if matchedHeight == nil {
+		nextHeight, _ := oldCursor.First()
+		return nextHeight
+	}
+
+	// Otherwise, find the next height to be migrated.
+	nextHeight, _ := oldCursor.Next()
+
+	return nextHeight
+}
+
+// locateNextUpdateNum returns a locator that's used to start our migration. A
+// nil locator means the migration has been finished.
+func locateNextUpdateNum(openChanBucket kvdb.RwBucket) (*updateLocator, error) {
+	locator := &updateLocator{}
+
+	// cb is the callback function to be used when iterating the buckets.
+	cb := func(chanBucket kvdb.RwBucket, l *updateLocator) error {
+		locator = l
+		updateNum := findNextMigrateHeight(chanBucket)
+
+		// We've found the next commit height and can now exit.
+		if updateNum != nil {
+			locator.nextHeight = updateNum
+			return errExit
+		}
+		return nil
+	}
+
+	// Iterate the buckets. If we received an exit signal, return the
+	// locator.
+	err := iterateBuckets(openChanBucket, nil, cb)
```

Comment:
Shouldn't we be passing in a specific node's bucket here rather than the entire open channel bucket? As we need to know the last location for each channel, not the last _global_ key we migrated?

Reply:
Actually it's designed to locate the global key. Maybe I should've named the functions better, but `findNextMigrateHeight` is designed to find the next key within the scope of a *channel*, and `locateNextUpdateNum` is to find the *global* position.

Suppose we restart a previously interrupted migration, we need to,
```
-> iterate the node's buckets until,
    -> iterate the chainhash buckets until,
        -> iterate the channel buckets until,
            -> a channel's next revocation log height is located, which signals the exit.
```

In other words a restart means we need to find out which channel inside which node has unfinished jobs and continue the migration there.